src/extensions/idle.rs

```diff
@@ -100,37 +140,65 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
     ///
     /// This is necessary so that we can keep using the inner `Session` in `wait_keepalive`.
-    fn wait_inner(&mut self, reconnect: bool) -> Result<WaitOutcome> {
+    fn wait_inner<F>(&mut self, reconnect: bool, mut callback: F) -> Result<WaitOutcome>
+    where
+        F: FnMut(UnsolicitedResponse) -> CallbackAction,
+    {
         let mut v = Vec::new();
-        loop {
-            let result = match self.session.readline(&mut v).map(|_| ()) {
+        let result = loop {
+            let rest = match self.session.readline(&mut v) {
                 Err(Error::Io(ref e))
                     if e.kind() == io::ErrorKind::TimedOut
                         || e.kind() == io::ErrorKind::WouldBlock =>
                 {
-                    if reconnect {
-                        self.terminate()?;
-                        self.init()?;
-                        return self.wait_inner(reconnect);
+                    break Ok(WaitOutcome::TimedOut);
+                }
+                Ok(_len) => {
+                    //  Handle Dovecot's imap_idle_notify_interval message
+                    if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+                        v.clear();
+                        continue;
+                    }
+                    match parse_idle(&v) {
+                        (_rest, Some(Err(r))) => break Err(r),
+                        (rest, Some(Ok(response))) => {
+                            if let CallbackAction::Stop = callback(response) {
+                                break Ok(WaitOutcome::MailboxChanged);
+                            }
+                            rest
+                        }
+                        (rest, None) => rest,
                     }
-                    Ok(WaitOutcome::TimedOut)
                 }
-                Ok(()) => Ok(WaitOutcome::MailboxChanged),
-                Err(r) => Err(r),
-            }?;
+                Err(r) => break Err(r),
+            };
 
-            // Handle Dovecot's imap_idle_notify_interval message
-            if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+            // Update remaining data with unparsed data if needed.
+            if rest.is_empty() {
                 v.clear();
-            } else {
-                break Ok(result);
+            } else if rest.len() != v.len() {
+                v = rest.into();
+            }
+        };
+
+        // Reconnect on timeout if needed
+        match (reconnect, result) {
+            (true, Ok(WaitOutcome::TimedOut)) => {
+                self.terminate()?;
+                self.init()?;
+                self.wait_inner(reconnect, callback)
             }
+            (_, result) => result,
         }
     }
 
-    /// Block until the selected mailbox changes.
-    pub fn wait(mut self) -> Result<()> {
-        self.wait_inner(true).map(|_| ())
+    /// Block until the given callback returns `Stop`, or until an unhandled
+    /// response arrives on the IDLE channel.
```

Comment:
What is an "unhandled response" here?

Reply:
Ah, looking at the parser, it seems like it's anything that we don't _explicitly_ recognize as a possible unsolicited response. That should be documented somewhere. I also wonder if we should instead have an `Unexpected` category under `UnsolicitedResponse` that just contains the raw `imap_proto::Response` — what do you think?

Reply:
Yes, it's just anything we do not recognize as an `UnsolicitedResponse`, so anything we haven't explicitly handled.

I am not sure if there will be ownership issues pushing a complete `imap_proto::Response` into the unsolicited responses `channel` - ownership issues are the primary reason why previous PRs for this functionality have wrapped the various `imap_proto` types, and I do not see `into_owned` on `imap_proto::Response` (but could just be blind).

I _do_ think it is kind of weird and confusing to have two types of "Response" in `UnsolicitedResponse` though - the kind we wrap, and the kind we don't. If we want to just give the user the `imap_proto::Response` directly, then why bother wrapping at all? Why not just push `imap_proto::Response` into the unsolicited responses `channel` all the time? This would be vastly simpler for us since we would not have to wrap things at all and all the ownership questions would be resolved in only one way, and it would be simpler for the user since there would only be one "Response" to match against.

Reply:
Re documentation: Do you just mean that what is meant by "unhandled response" should be more explicit? I can fix up the language there.

If you mean that it isn't clear that `UnsolicitedResponse` is incomplete, then where would be the best place to make that clear? The documentation for the `UnsolicitedResponse` type says "these are the ones that we have encountered in the wild", which _implies_ incompleteness, but doesn't come right out and say it. The documentation for `TryFrom<Response<'a>> for UnsolicitedResponse` explicitly says that not all possible `Response` variants are supported.

Reply:
Yeah, I'm torn here. I think you're right that `Unexpected` is a bit of a cop-out. I was thinking of it as a "these are the responses you might want to handle, and here's an escape hatch for if there's something we didn't think of", but maybe that escape hatch doesn't buy us much. Plus, it means that we're more likely to hear about missed variants, which is a plus. I guess then let's stick with dropping unexpected unsolicited responses, and expect `UnsolicitedResponse` to grow over time instead.

And yes, I think documentation-wise this should be updated to refer more clearly to what "unhandled response" means. I _also_ think it'd be good to be even more explicit in the docs for `UnsolicitedResponse` ­— maybe even link to the issue tracker and suggest that people file an issue if they see an unsolicited response that isn't captured by the enum — what do you think?

Reply:
I like the idea of just making it explicit that `UnsolicitedResponse` is only covering things we actually expect to see, and adding a link to the issue tracker if someone finds something new. I changed the language a bit on `UnsolicitedResponse` to this effect, and then changed the language on the `wait_*` functions to be more clear and linked them to `UnsolicitedResponse`.